File: packages/shared/src/shared/http/tasks.py
from __future__ import annotations

from datetime import datetime

from pydantic import Field, JsonValue

from shared.deployments import StubKind
from shared.http.base import HttpModel
from shared.http.compute import ContainerResponse
from shared.http.task_progress import TaskPendingProgress
from shared.tasks import TaskStatus


class TaskActionCapabilitiesResponse(HttpModel):
    can_cancel: bool = False
    can_rerun: bool = False
    can_shell: bool = False


class TaskAppReferenceResponse(HttpModel):
    name: str


class TaskWorkloadReferenceResponse(HttpModel):
    name: str
    kind: StubKind


class TaskDeploymentReferenceResponse(HttpModel):
    name: str
    version: int


class TaskSummaryResponse(HttpModel):
    """Task status and resource references, without execution inputs or results."""

    id: str
    name: str
    status: TaskStatus = TaskStatus.Pending
    pending_progress: TaskPendingProgress | None = None
    workspace_id: str | None = None
    app_id: str | None = None
    stub_id: str | None = None
    deployment_id: str | None = None
    container_id: str | None = None
    parent_task_id: str | None = None
    root_task_id: str | None = None
    handler: str | None = None
    attempt_number: int = 0
    max_attempts: int = 1
    next_retry_at: datetime | None = None
    exit_code: int | None = None
    created_at: datetime
    started_at: datetime | None = None
    finished_at: datetime | None = None
    app: TaskAppReferenceResponse | None = None
    workload: TaskWorkloadReferenceResponse | None = None
    deployment: TaskDeploymentReferenceResponse | None = None
    actions: TaskActionCapabilitiesResponse = Field(default_factory=TaskActionCapabilitiesResponse)


class TaskResponse(TaskSummaryResponse):
    command: list[str] = Field(default_factory=list)
    args: list[JsonValue] = Field(default_factory=list)
    kwargs: dict[str, JsonValue] = Field(default_factory=dict)
    result: JsonValue = None
    error: str | None = None


class TaskDetailResponse(TaskResponse):
    """Task inputs, result and the container it ran in."""

    container: ContainerResponse | None = None


class TaskPageResponse(HttpModel):
    data: list[TaskSummaryResponse] = Field(default_factory=list)
    next: str = ""


class TaskCountByDeploymentResponse(HttpModel):
    deployment_id: str
    count: int
    status_counts: dict[TaskStatus, int] = Field(default_factory=dict)


class TaskCountByDeploymentListResponse(HttpModel):
    items: list[TaskCountByDeploymentResponse] = Field(default_factory=list)


class TaskTimeWindowBucketResponse(HttpModel):
    timestamp: datetime
    count: int
    status_counts: dict[TaskStatus, int] = Field(default_factory=dict)


class TaskTimeWindowBucketListResponse(HttpModel):
    items: list[TaskTimeWindowBucketResponse] = Field(default_factory=list)


class TaskStopResponse(HttpModel):
    stopped: list[str] = Field(default_factory=list)
    skipped: list[str] = Field(default_factory=list)


class TaskMetricsSummaryResponse(HttpModel):
    total: int
    status_counts: dict[TaskStatus, int] = Field(default_factory=dict)
    completed: int
    failed: int
    cancelled: int
    failure_rate: float = 0.0
    average_runtime_ms: float | None = None
    runtime_ms_p50: float | None = None
    runtime_ms_p95: float | None = None
    runtime_ms_p99: float | None = None
    startup_ms_p50: float | None = None
    startup_ms_p95: float | None = None


__all__ = [
    "TaskActionCapabilitiesResponse",
    "TaskAppReferenceResponse",
    "TaskCountByDeploymentListResponse",
    "TaskCountByDeploymentResponse",
    "TaskDeploymentReferenceResponse",
    "TaskDetailResponse",
    "TaskMetricsSummaryResponse",
    "TaskPageResponse",
    "TaskResponse",
    "TaskStopResponse",
    "TaskSummaryResponse",
    "TaskTimeWindowBucketListResponse",
    "TaskTimeWindowBucketResponse",
    "TaskWorkloadReferenceResponse",
]

File: apps/web/src/lib/api/schemas/tasks.ts
import { z } from "zod";

import { containerSchema } from "./compute";
import { jsonValueSchema } from "./json";
import { taskPendingProgressSchema } from "./task-progress";

// Synced to packages/shared/src/shared/http/tasks.py (TaskResponse, TaskDetailResponse,
// TaskPageResponse, TaskMetricsSummaryResponse, TaskTimeWindowBucketListResponse) and
// packages/shared/src/shared/http/functions.py (FunctionCallGraphNode, FunctionCallGraphResponse).

export const taskStatuses = [
  "pending",
  "running",
  "retry",
  "complete",
  "failed",
  "expired",
  "timeout",
  "cancelled",
] as const;
export type TaskStatusValue = (typeof taskStatuses)[number];

const terminalTaskStatuses: readonly TaskStatusValue[] = [
  "complete",
  "failed",
  "expired",
  "timeout",
  "cancelled",
];

/**
 * The owning resources a task row names.
 *
 * The server sends what a reader sees — a name, a kind, a version — rather than
 * the whole app, workload, and deployment records, which a page of a hundred
 * rows would otherwise repeat a hundred times. Their ids are on the row itself.
 */
const taskAppReferenceSchema = z.object({ name: z.string() });
const taskWorkloadReferenceSchema = z.object({ name: z.string(), kind: z.string() });
const taskDeploymentReferenceSchema = z.object({ name: z.string(), version: z.number() });

export const taskSchema = z.object({
  id: z.string(),
  name: z.string(),
  status: z.string(),
  pending_progress: taskPendingProgressSchema.nullish(),
  workspace_id: z.string().nullish(),
  app_id: z.string().nullish(),
  stub_id: z.string().nullish(),
  deployment_id: z.string().nullish(),
  container_id: z.string().nullish(),
  parent_task_id: z.string().nullish(),
  root_task_id: z.string().nullish(),
  handler: z.string().nullish(),
  attempt_number: z.number().default(0),
  max_attempts: z.number().default(1),
  result: jsonValueSchema.nullish(),
  error: z.string().nullish(),
  exit_code: z.number().nullish(),
  created_at: z.string(),
  started_at: z.string().nullish(),
  finished_at: z.string().nullish(),
  app: taskAppReferenceSchema.nullish(),
  workload: taskWorkloadReferenceSchema.nullish(),
  deployment: taskDeploymentReferenceSchema.nullish(),
  // Only the single-task read (TaskDetailResponse) resolves a container; rows
  // carry `container_id` and nothing more.
  container: containerSchema.nullish(),
  actions: z
    .object({
      can_cancel: z.boolean().default(false),
      can_rerun: z.boolean().default(false),
      can_shell: z.boolean().default(false),
    })
    .default({ can_cancel: false, can_rerun: false, can_shell: false }),
});
export type Task = z.infer<typeof taskSchema>;

export const taskPageSchema = z.object({
  data: z.array(taskSchema).default([]),
  next: z.string().default(""),
});

export const taskMetricsSummarySchema = z.object({
  total: z.number(),
  status_counts: z.record(z.number()).default({}),
  completed: z.number(),
  failed: z.number(),
  cancelled: z.number(),
  failure_rate: z.number().default(0),
  average_runtime_ms: z.number().nullish(),
  runtime_ms_p50: z.number().nullish(),
  runtime_ms_p95: z.number().nullish(),
  runtime_ms_p99: z.number().nullish(),
  startup_ms_p50: z.number().nullish(),
  startup_ms_p95: z.number().nullish(),
});
export type TaskMetricsSummary = z.infer<typeof taskMetricsSummarySchema>;

const taskTimeWindowBucketSchema = z.object({
  timestamp: z.string(),
  count: z.number(),
  status_counts: z.record(z.number()).default({}),
});
export type TaskTimeWindowBucket = z.infer<typeof taskTimeWindowBucketSchema>;

export const taskTimeWindowBucketListSchema = z.object({
  items: z.array(taskTimeWindowBucketSchema).default([]),
});

export function isKnownTaskStatus(status: string): status is TaskStatusValue {
  return (taskStatuses as readonly string[]).includes(status);
}

export function isTerminalTaskStatus(status: string): boolean {
  return (terminalTaskStatuses as readonly string[]).includes(status);
}

const callGraphNodeBaseSchema = z.object({
  task_id: z.string(),
  parent_task_id: z.string().default(""),
  root_task_id: z.string().default(""),
  status: z.string(),
  name: z.string().default(""),
  function_name: z.string().default(""),
  created_at: z.string().nullish(),
  started_at: z.string().nullish(),
  finished_at: z.string().nullish(),
  dependencies: z.array(z.string()).default([]),
});

export type CallGraphNode = z.infer<typeof callGraphNodeBaseSchema> & {
  children: CallGraphNode[];
};

const callGraphNodeSchema: z.ZodType<CallGraphNode, z.ZodTypeDef, unknown> =
  callGraphNodeBaseSchema.extend({
    children: z.lazy(() => z.array(callGraphNodeSchema).default([])),
  });

export const callGraphSchema = z.object({
  root_task_id: z.string().default(""),
  root: callGraphNodeSchema.nullish(),
  nodes: z.array(callGraphNodeSchema).default([]),
});
export type CallGraph = z.infer<typeof callGraphSchema>;
